File: my-messenger-backend/my.messenger.common/Messaging/MessageService.cs
using my.messenger.common.Users;
using System;
using System.Collections.Generic;
using System.Text;
using System.Threading.Tasks;

namespace my.messenger.common.Messaging
{
    public class MessageService
    {
        private IUserRepository userRepository;
        private IMessageSender messageSender;
        private IMessageReceiver messageReceiver;

        public MessageService(IUserRepository userRepository, IMessageSender messageSender, IMessageReceiver messageReceiver)
        {
            this.userRepository = userRepository;
            this.messageSender = messageSender;
            this.messageReceiver = messageReceiver;
        }

        public ISubscriber Subscribe(String userId, IMessageConsumer consumer)
        {
            return messageReceiver.Subscribe(userId, consumer);
        }

        public Task<List<Message>> Dequeue(string userId, int count)
        {
            return messageReceiver.Dequeue(userId, count);
        }

        public async Task SendAsync(Message message)
        {
            if (!StringUtils.hasText(message.FromUserId))
            {
                throw new MessageServiceException("Invalid source");
            }

            if (message.To == null || !StringUtils.hasText(message.To.Id))
            {
                throw new MessageServiceException("Invalid source");
            }

            if (String.Equals(message.To.Id, message.FromUserId, StringComparison.InvariantCultureIgnoreCase))
            {
                throw new MessageServiceException("source equals destination");
            }

            if (!StringUtils.hasText(message.Body))
            {
                throw new MessageServiceException("Invalid body");
            }

            if (message.To.Type == DestinationType.User)
            {
                if (await userRepository.FindByIdAsync(message.To.Id) == null)
                {
                    throw new MessageServiceException("Destination User Not Found");
                }
            }

            await messageSender.SendAync(message);
        }
    }
}

File: my-messenger-backend-java/src/main/java/mymessenger/backend/services/MessageService.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package mymessenger.backend.services;

import mymessenger.backend.messaging.Subscriber;
import java.util.logging.Level;
import java.util.logging.Logger;
import mymessenger.backend.messaging.MessageConsumer;
import mymessenger.backend.messaging.MessageSender;
import mymessenger.backend.messaging.MessageSenderException;
import mymessenger.backend.model.messaging.DestinationType;
import mymessenger.backend.model.messaging.Message;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.util.StringUtils;
import mymessenger.backend.messaging.MessageReceiver;

/**
 *
 * @author guilherme
 */
@Service
public class MessageService {

    private final MessageSender messageSender;
    private final UserRepository userRepository;
    private final MessageReceiver messageReceiver;

    public MessageService(
            @Autowired MessageSender messageSender,
            @Autowired UserRepository userRepository,
            @Autowired MessageReceiver listenerService) {
        this.messageSender = messageSender;
        this.userRepository = userRepository;
        this.messageReceiver = listenerService;
    }

    public Subscriber subscribe(String userId, MessageConsumer consumer) throws MessageServiceException {
        try {
            return messageReceiver.subscribe(userId, consumer);
        } catch (MessageSenderException ex) {
            throw new MessageServiceException("Unable to listen", ex);
        }
    }

    public void send(Message message) throws MessageServiceException {

        if (!StringUtils.hasText(message.getFromUserId())) {
            throw new MessageServiceException("Invalid source");
        }

        if (message.getTo() == null || !StringUtils.hasText(message.getTo().getId())) {
            throw new MessageServiceException("Invalid source");
        }

        if (!StringUtils.hasText(message.getBody())) {
            throw new MessageServiceException("Invalid body");
        }

        if (message.getType() == null) {
            throw new MessageServiceException("Invalid Type");
        }

        if (message.getTo().getType() == DestinationType.User) {
            if (!userRepository.findById(message.getTo().getId()).isPresent()) {
                throw new MessageServiceException("Destination User Not Found");
            }
        }

        try {
            messageSender.send(message);
        } catch (MessageSenderException ex) {
            throw new MessageServiceException(ex);
        }
    }
}
